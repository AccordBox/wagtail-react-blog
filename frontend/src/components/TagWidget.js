import React from "react";
import axios from "axios";
import { Link } from "react-router-dom";

class TagWidget extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      tags: [],
      loading: true,
    };
  }

  componentDidMount() {
    axios.get("/api/blog/tags/").then((res) => {
      const tags = res.data.results;
      this.setState({
        tags,
        loading: false
      });
    });
  }

  render() {
    let content;
    if (this.state.loading) {
      content = 'Loading...';
    } else {
      content = this.state.tags.map((tag) => (
        <Link to={`/tag/${tag.slug}`} key={tag.slug}>
          <span className="badge badge-secondary">{tag.name}</span>{" "}
        </Link>
      ))
    }

    return (
      <div className="card my-4">
        <h5 className="card-header">Tags</h5>
        <div className="card-body">
          {content}
        </div>
      </div>
    );
  }
}

export { TagWidget };
