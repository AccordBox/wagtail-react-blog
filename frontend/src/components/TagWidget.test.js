import React from "react";
import { render, screen, wait} from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import axios from 'axios';

import { TagWidget } from "./TagWidget";

jest.mock('axios');

test('render Tag widget', async () => {
  const resp = {
    data: {
      results: [
        {
          slug: "wagtail",
          name: "Wagtail",
        },
        {
          slug: "django",
          name: "Django",
        },
        {
          slug: "react",
          name: "React",
        },
      ],
    }
  };
  axios.get.mockResolvedValue(resp);

  const { asFragment } = render(
    <MemoryRouter>
      <TagWidget />
    </MemoryRouter>
  );
  expect(screen.getByText("Loading...")).toBeInTheDocument();

  await wait(() => expect(axios.get).toHaveBeenCalled());

  await wait(() => expect(screen.getByText("Wagtail")).toBeInTheDocument());
  const el = screen.getByText("Wagtail");
  expect(el.tagName).toEqual('SPAN');
  expect(el).toHaveClass('badge badge-secondary');

  resp.data.results.map((tag) =>
    expect(screen.getByText(tag.name)).toBeInTheDocument()
  );

  expect(asFragment()).toMatchSnapshot();

});
