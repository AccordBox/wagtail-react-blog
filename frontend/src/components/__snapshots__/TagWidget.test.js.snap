// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Tag widget 1`] = `
<DocumentFragment>
  <div
    class="card my-4"
  >
    <h5
      class="card-header"
    >
      Tags
    </h5>
    <div
      class="card-body"
    >
      <a
        href="/tag/wagtail"
      >
        <span
          class="badge badge-secondary"
        >
          Wagtail
        </span>
         
      </a>
      <a
        href="/tag/django"
      >
        <span
          class="badge badge-secondary"
        >
          Django
        </span>
         
      </a>
      <a
        href="/tag/react"
      >
        <span
          class="badge badge-secondary"
        >
          React
        </span>
         
      </a>
    </div>
  </div>
</DocumentFragment>
`;
